{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fed3dafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import os\n",
    "import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "957b777a",
   "metadata": {},
   "outputs": [],
   "source": [
    "DIR = \"/home/yutianc/sars-cov-2-dms-ref\"\n",
    "DATA_DIR = \"/home/yutianc/sars-cov-2-dms-ref/data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8a280c8a",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/yutianc/sars-cov-2-dms-ref/output/escape_all.fa'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m RBD_COORDS \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m22553\u001b[39m, \u001b[38;5;241m23155\u001b[39m] \u001b[38;5;66;03m# 1-based idx\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m records \u001b[38;5;241m=\u001b[39m \u001b[43mAlignIO\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpath\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43mDIR\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43moutput/escape_all.fa\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m rec \u001b[38;5;129;01min\u001b[39;00m records:\n\u001b[1;32m      6\u001b[0m     seq_id \u001b[38;5;241m=\u001b[39m rec\u001b[38;5;241m.\u001b[39mid\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\u001b[38;5;241m.\u001b[39mlower()\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/Bio/AlignIO/__init__.py:384\u001b[0m, in \u001b[0;36mread\u001b[0;34m(handle, format, seq_count)\u001b[0m\n\u001b[1;32m    382\u001b[0m iterator \u001b[38;5;241m=\u001b[39m parse(handle, \u001b[38;5;28mformat\u001b[39m, seq_count)\n\u001b[1;32m    383\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 384\u001b[0m     alignment \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43miterator\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    385\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m:\n\u001b[1;32m    386\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo records found in handle\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/Bio/AlignIO/__init__.py:323\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(handle, format, seq_count)\u001b[0m\n\u001b[1;32m    320\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m seq_count \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(seq_count, \u001b[38;5;28mint\u001b[39m):\n\u001b[1;32m    321\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNeed integer for seq_count (sequences per alignment)\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 323\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m as_handle(handle) \u001b[38;5;28;01mas\u001b[39;00m fp:\n\u001b[1;32m    324\u001b[0m     \u001b[38;5;66;03m# Map the file format to a sequence iterator:\u001b[39;00m\n\u001b[1;32m    325\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mformat\u001b[39m \u001b[38;5;129;01min\u001b[39;00m _FormatToIterator:\n\u001b[1;32m    326\u001b[0m         iterator_generator \u001b[38;5;241m=\u001b[39m _FormatToIterator[\u001b[38;5;28mformat\u001b[39m]\n",
      "File \u001b[0;32m~/miniforge3/envs/env/lib/python3.9/contextlib.py:119\u001b[0m, in \u001b[0;36m_GeneratorContextManager.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    117\u001b[0m \u001b[38;5;28;01mdel\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mkwds, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfunc\n\u001b[1;32m    118\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 119\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mnext\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgen\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    120\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mStopIteration\u001b[39;00m:\n\u001b[1;32m    121\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenerator didn\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt yield\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/Bio/File.py:72\u001b[0m, in \u001b[0;36mas_handle\u001b[0;34m(handleish, mode, **kwargs)\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Context manager to ensure we are using a handle.\u001b[39;00m\n\u001b[1;32m     31\u001b[0m \n\u001b[1;32m     32\u001b[0m \u001b[38;5;124;03mContext manager for arguments that can be passed to SeqIO and AlignIO read, write,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     69\u001b[0m \n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 72\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mhandleish\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m fp:\n\u001b[1;32m     73\u001b[0m         \u001b[38;5;28;01myield\u001b[39;00m fp\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/yutianc/sars-cov-2-dms-ref/output/escape_all.fa'"
     ]
    }
   ],
   "source": [
    "RBD_COORDS = [22553, 23155] # 1-based idx\n",
    "\n",
    "records = AlignIO.read(os.path.join(DIR, \"output/escape_all.fa\"), format=\"fasta\")\n",
    "\n",
    "for rec in records:\n",
    "    seq_id = rec.id.split(\"_\")[-2].lower()\n",
    "\n",
    "    mut_file = glob.glob(os.path.join(DATA_DIR, seq_id, \"mut1*.csv\"))[0]\n",
    "    mut_df = pd.read_csv(mut_file)\n",
    "    mut_wt = list(mut_df[\"wt\"])\n",
    "    wt = []\n",
    "    coords = []\n",
    "    start = 331\n",
    "\n",
    "    for pos in range(22553-1, 23155, 3):\n",
    "        try:\n",
    "            aa = str(Seq(rec.seq[pos: pos+3]).translate())\n",
    "            wt.append(aa)\n",
    "            coords.append(start) # return 1-based idx\n",
    "            start += 1\n",
    "        except:\n",
    "            start += 1\n",
    "            continue\n",
    "\n",
    "            \n",
    "    # df = pd.DataFrame({\"wt\": wt, \"pos\": coords})\n",
    "    df = pd.DataFrame({\"wt\": wt, \"mut_wt\": mut_wt, \"coords\": coords})\n",
    "    df_diff = df.loc[df[\"wt\"] != df[\"mut_wt\"], :]\n",
    "    print(df_diff)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e652b5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7da72130",
   "metadata": {},
   "outputs": [],
   "source": [
    "RBD_COORDS = [22553, 23155] # 1-based idx\n",
    "\n",
    "rec = SeqIO.read(\"/home/yutianc/bjorn_rep/data/sc2/NC_045512.2_BA.1.fasta\", format=\"fasta\")\n",
    "\n",
    "mut_file = \"/home/yutianc/sars-cov-2-dms-ref/data/ba.1/mut1_BA.1_Omicron_baseline_EPI_ISL_10000028.csv\"\n",
    "mut_df = pd.read_csv(mut_file)\n",
    "mut_wt = list(mut_df[\"wt\"])\n",
    "\n",
    "wt = []\n",
    "coords = []\n",
    "start = 331\n",
    "\n",
    "for pos in range(22553-1, 23155, 3):\n",
    "    try:\n",
    "        aa = str(Seq(rec.seq[pos: pos+3]).translate())\n",
    "        wt.append(aa)\n",
    "        coords.append(start) # return 1-based idx\n",
    "        start += 1\n",
    "    except:\n",
    "        start += 1\n",
    "        continue\n",
    "\n",
    "\n",
    "df = pd.DataFrame({\"wt\": wt, \"mut_wt\": mut_wt, \"coords\": coords})\n",
    "df_diff = df.loc[df[\"wt\"] != df[\"mut_wt\"], :]\n",
    "print(df_diff)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e9549b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "FA_EXTS = {\".fa\", \".fasta\", \".fna\"}\n",
    "\n",
    "def iter_fasta_files(root: str):\n",
    "    root_path = Path(root)\n",
    "    for p in root_path.rglob(\"*\"):\n",
    "        if p.is_file() and p.suffix.lower() in FA_EXTS:\n",
    "            yield p\n",
    "\n",
    "\n",
    "def extract_ids_from_fasta_file(fp: Path):\n",
    "    ids = []\n",
    "    fname = []\n",
    "    with fp.open(\"r\", encoding=\"utf-8\", errors=\"replace\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                # first token after '>' (drops description)\n",
    "                seq_id = line[1:].strip().split()[0]\n",
    "                if seq_id:\n",
    "                    try:\n",
    "                        sra = seq_id.split('/')[2]\n",
    "                        fa_name = (f.name.split(\"/\")[-1]).split('.')[0]\n",
    "                        if sra != fa_name:\n",
    "                            ids.append(sra)\n",
    "                            fname.append(fa_name)\n",
    "                    except:\n",
    "                        continue\n",
    "    return ids, fname\n",
    "\n",
    "def extract_ids_from_dir(root: str):\n",
    "    out_id, out_fname = [], []\n",
    "    for fp in iter_fasta_files(root):\n",
    "        ids, fname = extract_ids_from_fasta_file(fp)\n",
    "        out_id.extend(ids)\n",
    "        out_fname.extend(fname)\n",
    "    print(len(out_fname))\n",
    "    print(len(out_id))\n",
    "\n",
    "    return pd.DataFrame({\"ID\": out_id, \"fname\": out_fname})\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    root_dir = \"/home/yutianc/bjorn_rep/data/sc2/consensus_sequences/\"\n",
    "    df = extract_ids_from_dir(root_dir)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffcfb60f",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = dict(zip(df[\"fname\"], df[\"ID\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c274fc0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81df96a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lineage_all = pd.read_csv(\"/home/yutianc/bjorn_rep/data/sc2/lineage_report.csv\")\n",
    "lineage_all[\"taxon\"] = lineage_all[\"taxon\"].map(d).fillna(lineage_all[\"taxon\"])\n",
    "lineage_all = lineage_all[[\"taxon\", \"lineage\"]]\n",
    "lineage_all = lineage_all.rename(columns={\"taxon\": \"sample\", \"lineage\": \"Genotype\"})\n",
    "lineage_all.to_csv(\"/home/yutianc/bjorn_rep/data/sc2/lineage_report_cleaned.tsv\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0600f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e23b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "root  = Path('/home/yutianc/bjorn_rep/data/sc2/consensus_sequences/')\n",
    "sra_all = [(f.name).split(\".\")[0] for f in root.rglob(\"*\") if f.is_file()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "718b1e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sra_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dd2e4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = pd.read_csv(\"/home/jmcfeeters/sc2_data/2025.06.12_outputs_sc2/metadata/normalized_metadata.tsv\", sep='\\t')\n",
    "meta = meta[:len(sra_all)]\n",
    "meta[\"Accession\"] = sra_all\n",
    "len(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36376186",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8e55530",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta.to_csv('/home/yutianc/muninn/inputs/metadata_sc2_all.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4cb26d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut = pd.read_csv(\"/home/yutianc/muninn/inputs/escape_all_1000/mutations.tsv\", sep='\\t')\n",
    "meta_sample = meta[meta[\"Accession\"].isin(list(mut[\"sra\"].unique()))]\n",
    "meta_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "553e5921",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut[\"sra\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f55ac92",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
